### PR TITLE
Added Typing

### LINKED ISSUE #301: Feature Request: Add Typing
I would be interested to see correct typing hints when using `xESMF`.

For my small function, I am getting an error at the return
```python
def regrid_to_lat_lon(
    ds: xr.Dataset,
    target_res_lat: float = 1,
    target_res_lon: float = 1,
) -> xr.Dataset:
    ds_out = xe.util.grid_global(target_res_lat, target_res_lon, cf=True)
    regridder = xe.Regridder(
        ds,
        ds_out,
        "bilinear",
        periodic=True,
    )
    dr_out = regridder(ds, keep_attrs=True)
    return dr_out
```
```
Expression of type "NDArray[Any] | Any" cannot be assigned to return type "Dataset"
  Type "NDArray[Any] | Any" cannot be assigned to type "Dataset"
    "NDArray[Any]" is incompatible with "Dataset"Pylance[reportGeneralTypeIssues](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportGeneralTypeIssues)
```
There are probably not a lot of functions or classes which need typing hints for the average user.
I would also be interested to open a Pull Request for this, if the developers want it.

### PR DESCRIPTION
**Pull request by Siddhesh-Agarwal**
- Formatted with `isort` followed by `black` formatter
- Added Typing
- Closes #301

### CODE REVIEW

File: xesmf/backend.py [hunk 1]
```diff
@@ -26,7 +26,7 @@
 import numpy.lib.recfunctions as nprec
```
Comment (charlesgauthier-udm) [inline xesmf/backend.py:27]:
```suggestion
from typing import Optional
```
See below

**Comment by huard** [comment]
Thanks! I've asked a colleague to provide a review.

File: xesmf/backend.py [hunk 2]
```diff
@@ -58,7 +58,13 @@ def warn_lat_range(lat):
 
 class Grid(ESMF.Grid):
     @classmethod
-    def from_xarray(cls, lon, lat, periodic=False, mask=None):
+    def from_xarray(
+        cls,
+        lon: np.ndarray[float, int],
+        lat: np.ndarray[float, int],
+        periodic: bool = False,
+        mask=None,
```
Comment (charlesgauthier-udm) [inline xesmf/backend.py:66]:
```suggestion
        mask: Optional[np.ndarray[float, int]] = None,
```
Suggested for consistency with other args of `from_xarray`

**Review by charlesgauthier-udm — COMMENTED** [review]
Thank you for the great work @Siddhesh-Agarwal ! After review, here are my suggestions. I do not have the most experience with typing so please feel free to provide your input to the suggested changes.

**Comment by Siddhesh-Agarwal** [comment]
@charlesgauthier-udm I have completed the changes mentioned in the previous review. I hope these are up to your expectations.

File: xesmf/frontend.py [hunk 3]
```diff
@@ -611,22 +641,27 @@ def regrid_array(self, indata, weights, skipna=False, na_thres=1.0, output_chunk
...
         return self.regrid_array(indata, self.weights.data, **kwargs)
 
     def regrid_dataarray(
-        self, dr_in, keep_attrs=False, skipna=False, na_thres=1.0, output_chunks=None
+        self,
+        dr_in: xr.DataArray,
+        keep_attrs: bool = False,
+        skipna: bool = False,
+        na_thres: float = 1.0,
+        output_chunks: Optional[Tuple[int, ...]] = None,
```
[the remark this answers is not in the document]
Comment (huard) [reply xesmf/frontend.py:692]: Has this been addressed ?

[the remark this answers is not in the document]
Comment (Siddhesh-Agarwal) [reply xesmf/frontend.py]: It wasn't addressed until now. My latest commit fixes it.

File: xesmf/frontend.py [hunk 4]
```diff
@@ -702,15 +744,17 @@ def _parse_xrinput(self, dr_in):
...
-            temp_horiz_dims = [s + '_new' for s in input_horiz_dims]
+            temp_horiz_dims: List[str] = [s + '_new' for s in input_horiz_dims]
 
         if self.sequence_in and not self.sequence_out:
             temp_horiz_dims = ['dummy_new'] + temp_horiz_dims
         return input_horiz_dims, temp_horiz_dims
 
-    def _format_xroutput(self, out, new_dims=None):
+    def _format_xroutput(
+        self, out: xr.DataArray | xr.Dataset, new_dims: List[str]
```
[the remark this answers is not in the document]
Comment (huard) [reply xesmf/frontend.py:795]: Please keep the None default.

File: xesmf/smm.py [hunk 5]
```diff
@@ -3,168 +3,188 @@
...
+        raise ValueError("ny_in * nx_in should equal to weights.shape[1]")
 
     if shape_out[0] * shape_out[1] != weights.shape[0]:
-        raise ValueError('ny_out * nx_out should equal to weights.shape[0]')
+        raise ValueError("ny_out * nx_out should equal to weights.shape[0]")
 
 
-def apply_weights(weights, indata, shape_in, shape_out):
+def apply_weights(
+    weights: npt.NDArray[Any],
```
Comment (huard) [inline xesmf/smm.py:174]:
```suggestion
    weights: sps.COO,
```

**Review by huard — APPROVED** [review]

**Comment by huard** [comment]
Please resolve the test failures then this is ready to merge.

**Comment by Siddhesh-Agarwal** [comment]
@huard, it is insane to have these many pre-commit hooks. Black, Isort, and Autopep8 seem fine but now prettier? I am sorry but I will have to leave this PR in this state.

**Comment by huard** [comment]
Hi @Siddhesh-Agarwal 

Once installed locally, pre-commit will run before each commit, so this is not something you're expected to worry about and do "by-hand". 

I'm now working to address the remaining issues. I realized that by pushing to this PR, I'm pushing to your master, which might interfere with your work. Please open a PR in a separate branch to avoid this in the future.

**Comment by huard** [comment]
There are a couple of failures still remaining, but I'm stumped about what might cause them.

**Pull request closed without merging**